src/api/eth_filter.rs

```diff
@@ -114,104 +97,96 @@ impl<T: Transport, R> BaseFilter<T, R> {
   }
 }
 
-impl<T: Transport, R> Drop for BaseFilter<T, R> {
-  fn drop(&mut self) {
-    let _ = self.uninstall_internal().wait();
-  }
-}
-
-/// Logs Filter
-pub struct LogsFilter<T: Transport> {
-  base: BaseFilter<T, Log>,
-}
-
-impl<T: Transport> FilterHandle<T> for LogsFilter<T> {
-  fn new(id: U256, transport: T) -> Self {
-    LogsFilter {
-      base: BaseFilter {
-        id: id,
-        transport: transport,
-        _output: PhantomData,
-      }
-    }
-  }
-}
-
-impl<T: Transport> LogsFilter<T> {
-  /// Get all logs matching the filter.
+impl<T: Transport> BaseFilter<T, LogsFilter> {
+  /// Returns future with all logs matching given filter
   pub fn logs(&self) -> CallResult<Vec<Log>, T::Out> {
     let id = helpers::serialize(&self.id);
     CallResult::new(self.transport.execute("eth_getFilterLogs", vec![id]))
   }
 }
 
-impl<T: Transport> Deref for LogsFilter<T> {
-  type Target = BaseFilter<T, Log>;
-
-  fn deref(&self) -> &Self::Target {
-    &self.base
+impl<T: Transport, F: FilterInterface> Drop for BaseFilter<T, F> {
+  fn drop(&mut self) {
+    let _ = self.uninstall_internal().wait();
   }
 }
 
-/// New blocks hashes filter.
-pub struct BlocksFilter<T: Transport> {
-  base: BaseFilter<T, H256>,
+/// Should be used to create new filter future
+pub fn create_filter<T: Transport, F: FilterInterface>(t: T, arg: Vec<rpc::Value>) -> CreateFilter<T, F> {
+  let future = CallResult::new(t.execute(F::constructor(), arg));
+  CreateFilter {
+    transport: Some(t),
+    interface: PhantomData,
+    future: future,
+  }
 }
 
-impl<T: Transport> FilterHandle<T> for BlocksFilter<T> {
-  fn new(id: U256, transport: T) -> Self {
-    BlocksFilter {
-      base: BaseFilter {
-        id: id,
-        transport: transport,
-        _output: PhantomData,
-      }
-    }
-  }
+/// Future which resolves with new filter
+pub struct CreateFilter<T: Transport, F: FilterInterface> {
+  transport: Option<T>,
+  interface: PhantomData<F>,
+  future: CallResult<U256, T::Out>,
 }
 
-impl<T: Transport> Deref for BlocksFilter<T> {
-  type Target = BaseFilter<T, H256>;
+impl<T, F> Future for CreateFilter<T, F> where
```

Comment:
Just wondering if we really need a separate `Future` that resolves to `BaseFilter`. Can't the `BaseFilter` be stateful? Basically there is nothing you can do except for `and_then` on that filter, cause you need to wait for it to be created.